json-schema-form.schema.json

```
@@ -0,0 +1,153 @@
+    },
+    "properties": {
+      "additionalProperties": { "$ref": "#" }
+    },
+    "if": { "$ref": "#" },
+    "then": { "$ref": "#" },
+    "else": { "$ref": "#" },
+    "allOf": { "$ref": "#/definitions/schemaArray" },
+    "anyOf": { "$ref": "#/definitions/schemaArray" },
+    "oneOf": { "$ref": "#/definitions/schemaArray" },
+    "not": { "$ref": "#" },
+    "contains": { "$ref": "#" },
+    "propertyNames": { "$ref": "#" },
+    "x-jsf-order": {
+      "description": "This keyword defines the order of fields for a given form or fieldset. It's placed at the schema root and inside of each fieldset.",
+      "$ref": "http://json-schema.org/draft-07/schema#/definitions/stringArray"
+    },
+    "x-jsf-errorMessage": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "required": { "type": "string" },
+        "minimum": { "type": "string" },
+        "minLength": { "type": "string" },
+        "maximum": { "type": "string" },
+        "maxLength": { "type": "string" },
+        "pattern": { "type": "string" },
+        "maxFileSize": { "type": "string" },
+        "accept": { "type": "string" }
+      }
+    },
+    "x-jsf-presentation": {
+      "type": "object",
+      "description": "Presentation overrides for the schema",
+      "additionalProperties": false,
+      "properties": {
+        "inputType": {
+          "description": "Input type for the schema",
+          "type": "string",
+          "oneOf": [
+            {
+              "const": "text",
+              "description": "Similar to the native HTML input with text type."
+            },
+            {
+              "const": "select",
+              "description": "Similar to the native HTML select element."
+            },
+            {
+              "const": "radio",
+              "description": "Similar to a native HTML input with radio type."
+            },
+            {
+              "const": "number",
+              "description": "Similar to the native HTML input with number type."
+            },
+            {
+              "const": "date",
+              "description": "Expects a value with format YYY-MM-DD."
+            },
+            {
+              "const": "checkbox",
+              "description": "Similar to the native HTML input with checkbox type."
+            },
+            {
+              "const": "email",
+              "description": "Similar to the native HTML input with email type."
+            },
+            {
+              "const": "file",
+              "description": "Similar to the native HTML input with file type."
+            },
+            {
+              "const": "fieldset",
+              "description": "Groups multiple Fields inside. Its expected value is a nested object."
+            },
+            {
+              "const": "group-array",
+              "description": "A list of inputs that can be repeated. Its expected value is an array."
+            }
```

todo: Following the example above with the "money field". If the dev wants to represent a money field, probably it would look like this:

```jsonc
{
  "properties": {
    "budget": {
      "title": "What's your budget?",
      "type": "number",
      "x-jsf-presentation": {
        "inputType": "money", // <-- a custom inputType!
        "currency": "EUR"
      }
    }
  }
}
```

This means that, although JSF tries to guess the best field's `inputType` based on each property schema (when `x-jsf-presenttion` is not declared), it also allows you to pass any arbitrary value you need in your UI Form.

In other words, your meta schema is missing a last `oneOf`:

```suggestion
            },
            {
              "type": "string",
              "description": "Any arbitrary custom inputType you might want to represent your UI Field."
            }
```

The `inputType` does not influence JSF internals in any way. It's just for a UI flavor. Gotcha? ☺️

---

Does that mean I can also customize things like `statement` and `statement.severity`?

And is there anything that we should mark as required?

---

> Does that mean I can also customize things like statement and statement.severity?

Yes, exactly. We could say that `statement` is our ([Remote](https://remote.com/)) opinionated [(re)presentation of statements](https://json-schema-form.vercel.app/?path=/story/demos-fields--field-with-statement) in a form. In a future version, we could find a way to [remove it from JSF internals](https://github.com/remoteoss/json-schema-form/blob/main/src/helpers.js#LL455C9-L455C29)

> And is there anything that we should mark as required?

Good question. The only required key is `properties` in the root, as it's supposed to be a form (list) of fields.

JSF works totally fine with the most basic shape for each json schema property. (you can test it on the [playground](https://json-schema-form.vercel.app/?path=/docs/playground--playground)). Here's an example:

```json
{
  "properties": {
    "username": {},
    "age": {}
  }
}
```